### PR TITLE
Gather LLVM PGO profiles from `rustc-perf` suite on real-world crates

### PR DESCRIPTION
**Pull request by Kobzol**
This PR expands the benchmark suite used to gather LLVM PGO profiles in CI from `libcore` to several real-world crates. I hand-picked a few crates, but the list is up for debate.

Previous results that we got from running `syn,cargo,serde` looked pretty [good](https://perf.rust-lang.org/compare.html?start=2631aeef823a9e16d31f999d3f07001e5fcc4b3d&end=abf097afa10dde1aa3d8a1d422724a46aab79bf3).

Running `libcore` + `rustc-perf` with some number of crates is repeated now (and for BOLT it will also be needed), so maybe we can extract it to a bash function?

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I think extracting the running of perf into a function is a good step to take, yeah. I would be interested in seeing if we get similar results from just Full benchmarks (rather than running the full set of incremental benchmarks as well).

**Comment by Kobzol** [comment]
Okay. Currently, the PR expanded the list of executed crates (so it's not the same as my experiment from the previous PR). Do you want to run perf for this or should I remove the incremental builds right away?

**Comment by Kobzol** [comment]
[Here](https://perf.rust-lang.org/compare.html?start=abf097afa10dde1aa3d8a1d422724a46aab79bf3&end=6513a04b37f3b91602d705d5a18fcb1bf5983559&stat=instructions%3Au) is a comparison of `syn,cargo,serde` vs `syn,cargo,serde,ripgrep,regex,clap-rs,hyper-2`.

There is a slight improvement, but we are also hitting diminishing returns at this point. Dist build took 1h 21 minutes, which is similar to previous CI times.

**Comment by Kobzol** [comment]
I created a bash function (it's my first one, so go easy on me please :D ) and changed it to only run `Full` runs for LLVM PGO.

**Comment by lqd** [comment]
Just as an over-fitting sanity check: it would be cool to have a short summary of the effect this has on some crates we don't track the benchmark suite. The perf runs at least show the benefits on the crates that are not explicitly profiled for PGO. 

I can help with that, let me know when this is ready to try.

File: src/ci/pgo.sh [hunk 1]
```diff
@@ -36,38 +83,8 @@ python3 ../x.py build --target=$PGO_HOST --host=$PGO_HOST \
...
-        /checkout/obj/build/$PGO_HOST/stage2/bin/rustc \
-        Test \
-        --builds Check,Debug,Opt \
-        --cargo /checkout/obj/build/$PGO_HOST/stage0/bin/cargo \
-        --runs All \
-        --include externs,ctfe-stress-4,inflate,cargo,token-stream-stress,match-stress-enum
-
-cd /checkout/obj
+gather_profiles "Check,Debug,Opt" "All" \
+  "externs,ctfe-stress-4,inflate,cargo,token-stream-stress,match-stress-enum"
```
Comment (nnethercote) [inline src/ci/pgo.sh:87]: It makes me nervous that we have a silent cross-repository dependency here. I.e. removing or renaming benchmarks in `rustc-perf` will break `rust`. But I guess that's a pre-existing issue...

Comment (Kobzol) [reply src/ci/pgo.sh:87]: CI uses a pinned version of `rustc-perf` (https://github.com/rust-lang/rust/blob/master/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile#L101), so this shouldn't happen unless we update it manually.

Comment (nnethercote) [reply src/ci/pgo.sh:87]: Good to know, thanks!

**Comment by Mark-Simulacrum** [comment]
@rust-timer build ca29313890801dc91b8e5937e78d5d3546557874

**Comment by Kobzol** [comment]
[Here](https://perf.rust-lang.org/compare.html?start=6513a04b37f3b91602d705d5a18fcb1bf5983559&end=ca29313890801dc91b8e5937e78d5d3546557874&stat=instructions%3Au) is a comparsion of `All` runs (full + all incr. modes) vs only `Full` runs.

It seems to be mostly the same for instruction counts, but there are some larger regressions in cycle counts (not sure how relevant that is). Since it doesn't seem like running also `incr` modes adds a lot of time to CI, I would suggest to keep all of the runs in CI. WDYT?

**Comment by Mark-Simulacrum** [comment]
This seems good to me. I think we can probably endlessly iterate on the benchmark selection, but this shows a solid improvement already and that seems like it should be good enough to me.

@bors r+

**Pull request merged**